Redis configuration with URL

maust opened the pull request:
As some platforms provide Redis configuration through environment variables as URL (e.g. redis://redistogo:44ec0bc04dd4a5afe77a649acee7a8f3@drum.redistogo.com:9092/ on Heroku) it would be nice to be able to configure it without manually creating an JedisConnectionFactory similar to spring.datasource.url.

So I introduced: spring.redis.url which will parse the URL and create the JedisConnectionFactory.

As of the current implementation spring.redis.url would overrule spring.redis.host, spring.redis.port and spring.redis.password. But this would be open for discussion.

Note: There is only a password for Redis and no user, therefore the user in the URL will be ignored.

pivotal-issuemaster commented:
@maust Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=7395)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-boot&pullRequestId=7395) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/faq) for frequently asked questions.

pivotal-issuemaster commented:
@maust Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-boot&pullRequestId=7395)!

mp911de commented on spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/data/redis/RedisAutoConfiguration.java:
```diff
@@ -91,10 +98,25 @@ public JedisConnectionFactory redisConnectionFactory()
+			if (StringUtils.hasText(this.properties.getUrl())) {
+				try {
+					URI redisURI = new URI(this.properties.getUrl());
+					factory.setHostName(redisURI.getHost());
```
It would make sense to distinguish between plaintext (`redis://…`) and SSL (`rediss://`) connections.

maust replied:
How would it make any difference as the JedisFactory didn't distinguish before? Is there any parameter I should set if it is rediss:// ?

mp911de replied:
SSL has no property in `RedisProperties` yet. You can set whether to use SSL with [`JedisConnectionFactory.setUseSsl(…)`](http://docs.spring.io/spring-data/data-redis/docs/1.8.0.M1/api/org/springframework/data/redis/connection/jedis/JedisConnectionFactory.html#setUseSsl-boolean-).

mp911de commented on spring-boot-docs/src/main/asciidoc/appendix-application-properties.adoc:
```diff
@@ -787,6 +787,7 @@ content into your application; rather pick only the properties that you need.
 	spring.redis.cluster.max-redirects= # Maximum number of redirects to follow when executing commands across the cluster.
 	spring.redis.cluster.nodes= # Comma-separated list of "host:port" pairs to bootstrap from.
 	spring.redis.database=0 # Database index used by the connection factory.
+	spring.redis.url=redis://user:password@example.com:6379 # Connection URL, will override host, port and password (user will be ignored)
```
The user part is confusing as Redis supports only a password for authentication.

maust replied:
I agree, on the other side at least on heroku it is somehow the standard as it is the standard way of encoding user & password for http basic auth. I don't know how it works on other environments where the URL is provided through environment variables.

So we will have the following options:
1. Confuse the user with redis://user:password@...
2. Change it to redis://password@...
3. Change the implementation to allow 1. & 2. 
Best experience would be probably 3. but would make the RedisAutoConfiguration.java a bit more complicated.

Let me know what you think and I can see if I do the changes the following days.

mp911de replied:
Leaving the user portion blank (of userinfo) is the recommended way how to represent a Redis URI with credentials for now (`redis://:password@…`). Some clients accept `redis://password@…` but I'd rather stay on the safe side here.

See also https://www.iana.org/assignments/uri-schemes/prov/redis

maust commented:
@mp911de I added support for `rediss://..` so that it enables SSL.

After looking into it a bit more I also added spring.redis.ssl, which will also enable SSL if set to true (even if only a `redis://..` url is provided, or configuration is done differently, e.g. through spring.redis.host)

PS: Build did not work as the master seems to be broken. Could you please trigger the build after master is green again (don't know if I can or where)? Thanks.

philwebb commented:
See AMQP URL parsing issue in #6424 to make sure we're not making the same mistakes.

maust commented:
@philwebb As I understand #6424, the inconsistency is in the getters of RabbitProperties. The getters of RedisProperties are plain getters without any logic, over here the logic to decide which values to use is inside RedisAutoConfiguration. 

The logic is basicly in RedisAutoConfiguration:
- url always overrules hostname, port and password
- if no url is set hostname, port (and password if set) is used
- ssl can be used to enable SSL support independent if url is used or not

Compared to RabbitProperties the getter cannot be used e.g. to easily retrieve the hostname if the configuration is done through url. Or would this be the expected behavior?

Am I missing something else?

wilkinsona commented:
@maust Sorry, @philwebb's comment wasn't aimed at you. It was just making a note for something we need to check when merging.

philwebb commented:
Sorry @maust, I was typing the comment whilst we were discussing things on our call and didn't do a very good job of explaining what it was about.

philwebb commented:
Thanks for the PR. I've made a few changes in commit 90eb58252e678f6ea37b321e9377ece254fc9473 if you'd like to review.

Thanks again!
